### PR TITLE
[filters] Prevent a filter from sending local reply and continue

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Commit Message: Prevent filter iteration on headers after sending a local reply and returning `FilterHeadersStatus::Continue`. 
Additional Description:
Before, the test hits this crash:
```
[ RUN      ] Protocols/ProtocolIntegrationTest.ContinueAfterLocalReply/IPv4_Http2Downstream_Http2Upstream
[2020-12-14 16:51:35.110][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x0
[2020-12-14 16:51:35.110][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-12-14 16:51:35.110][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: 0/1.17.0-dev/test/RELEASE/BoringSSL
[2020-12-14 16:51:35.111][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: __restore_rt [0x7fc14cdc68a0]
[2020-12-14 16:51:35.115][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: Envoy::ConnectionPool::ConnPoolImplBase::attachStreamToClient() [0xc378d6]
[2020-12-14 16:51:35.120][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: Envoy::ConnectionPool::ConnPoolImplBase::onUpstreamReady() [0xc3824d]
[2020-12-14 16:51:35.124][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: Envoy::ConnectionPool::ConnPoolImplBase::onConnectionEvent() [0xc39118]
[2020-12-14 16:51:35.128][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: Envoy::Network::ConnectionImplBase::raiseConnectionEvent() [0x10eda1b]
[2020-12-14 16:51:35.132][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: Envoy::Network::ConnectionImpl::raiseEvent() [0x10e4889]
[2020-12-14 16:51:35.136][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Network::ConnectionImpl::onWriteReady() [0x10e744e]
[2020-12-14 16:51:35.140][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Network::ConnectionImpl::onFileEvent() [0x10e64e9]
[2020-12-14 16:51:35.144][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: std::__1::__function::__func<>::operator()() [0x10d0971]
[2020-12-14 16:51:35.148][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x10d174c]
[2020-12-14 16:51:35.152][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: event_process_active_single_queue [0x17c33f8]
[2020-12-14 16:51:35.156][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #11: event_base_loop [0x17c1dce]
[2020-12-14 16:51:35.160][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #12: Envoy::Server::WorkerImpl::threadRoutine() [0xac9688]
[2020-12-14 16:51:35.164][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #13: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x17bb2a3]
[2020-12-14 16:51:35.164][29][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #14: start_thread [0x7fc14cdbb6db]
```

The downside here is `iteration_state_` isn't modified to stop iteration. I don't think this matters since it's changed to `StopSingleIteration` in `maybeContinueDecoding` right after.

Risk Level: Low/medium in case this changes expected filter iteration?
Testing: Added test that shows message logs and behavior passes.
Fixes part of https://github.com/envoyproxy/envoy/issues/13737

### CODE REVIEW

**Comment by alyssawilk** [comment]
Thanks for the fix!  mind checking out CI and pinging back when it's sorted?

**Review by alyssawilk — APPROVED** [review]

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -447,22 +447,17 @@ void FilterManager::decodeHeaders(ActiveStreamDecoderFilter* filter, RequestHead
...
-    // A filter should not continue after a local reply was sent. Indicate the error and stop
-    // iteration.
-    if ((*entry)->complete() && status != FilterHeadersStatus::StopIteration) {
-      ENVOY_BUG(
-          false,
-          "Filters should not return FilterHeadersStatus::Continue after sending a local reply.");
-      status = FilterHeadersStatus::StopIteration;
-    }
+    ENVOY_BUG(!(*entry)->complete() || status == FilterHeadersStatus::StopIteration,
+              "Filters should not FilterHeadersStatus::StopIteration after sending a local reply.");
```
[the remark this answers is not in the document]
Comment (asraa) [reply source/common/http/filter_manager.cc]: Yes :) Thanks for catching this -- after iterating on how I should phrase it it ended up going backwards!

File: test/integration/protocol_integration_test.cc [hunk 2]
```diff
@@ -287,6 +288,31 @@ TEST_P(ProtocolIntegrationTest, ContinueHeadersOnlyInjectBodyFilter) {
...
+  // Send a headers only request.
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+  EXPECT_LOG_CONTAINS("error",
+                      "envoy bug failure: !(state_.local_complete_ ) || status == "
+                      "FilterHeadersStatus::StopIteration. Details: Filters should not "
+                      "FilterHeadersStatus::StopIteration after sending a local reply.",
+                      response->waitForEndStream());
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+#endif
```
[the remark this answers is not in the document]
Comment (asraa) [reply test/integration/protocol_integration_test.cc:316]: I'll add one and replace in here! https://github.com/envoyproxy/envoy/issues/14441

**Comment by asraa** [comment]
Just to double check, oauth2 filter sends a local reply and then sends StopAllIterationAndBuffer (the test failure). Is it possible that people can send a local reply and then continue header iteration?

The most minimal constraint would be changing the condition to "don't return continue after local reply" rather than "must return stopiteration after local reply"

**Comment by alyssawilk** [comment]
I wonder if it's an attempt to drain the connection so it can be reused (for HTTP/1.1 etc.)?
cc @rgs1 @derekargueta @snowp

**Comment by rgs1** [comment]
> Just to double check, oauth2 filter sends a local reply and then sends StopAllIterationAndBuffer (the test failure). Is it possible that people can send a local reply and then continue header iteration?
> 
> The most minimal constraint would be changing the condition to "don't return continue after local reply" rather than "must return stopiteration after local reply"

Thanks for this fix! I somehow forgot to follow-up with a fix here when we first bumped into this internally... this should be `StopIteration`. Our internal copy of this filter -- and similar filters -- have already being fixed... And that made the crash go away.

**Comment by rgs1** [comment]
> I wonder if it's an attempt to drain the connection so it can be reused (for HTTP/1.1 etc.)?
> cc @rgs1 @derekargueta @snowp

Which part? The Stop and Buffer as opposed to full Stop? Or continuing header iteration post local reply...? Afaik this was just a bug/omission on our side, not something to reset the connection state...

**Comment by alyssawilk** [comment]
Ah cool.  If it's just broken I'd be inclined to restore your last push and fix oath upstream too, WDYT?

**Review by rgs1 — APPROVED** [review]
Oauth bits look good.. Thank you!

File: source/extensions/filters/http/oauth2/filter.cc [hunk 3]
```diff
@@ -308,7 +308,7 @@ Http::FilterHeadersStatus OAuth2Filter::decodeHeaders(Http::RequestHeaderMap& he
   Http::Utility::Url state_url;
   if (!state_url.initialize(state_, false)) {
     sendUnauthorizedResponse();
-    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
```
Comment (alyssawilk) [inline source/extensions/filters/http/oauth2/filter.cc:311]: So I get switching the returns where we're sending a local reply, but it looks like you're switching from "stop all" to "stop" in other cases as well which I'm less confident of.

I thought oath was supposed to stop all progress until oath responded?  I think we need to revert those parts of the change.
FWIW I'd also be fine allowing any of the Stop calls if you think that's cleaner than fixing up specific returns of oath.

Comment (rgs1) [reply source/extensions/filters/http/oauth2/filter.cc:311]: Oh i missed this one:

```
  // pause while we await the next step from the OAuth server
-  return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
 + return Http::FilterHeadersStatus::StopIteration;
```

yeah I agree that one is fine to buffer I guess... but probably not a biggie if we fully pause until we hear back.

Comment (asraa) [reply source/extensions/filters/http/oauth2/filter.cc:311]: Sounds good, I'm not sure which makes sense -- I'll make the change less restrictive.

**Comment by asraa** [comment]
Sorry about the back and forth! All set with just preventing continue after a direct response. No changes to oauth2 (which maybe can be a TODO)

**Review by alyssawilk — APPROVED** [review]

File: test/integration/protocol_integration_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/integration/protocol_integration_test.cc:316]: Can you add a TODO about this? I agree we should have testing on the bug case.

/wait

**Comment by asraa** [comment]
Not sure if the wait is to update `filter_manager` to restrict back to stopiteration after local reply (rather than just "not continue"). Tests pass with @rgs1's Oauth PR merged.

File: test/integration/protocol_integration_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/integration/protocol_integration_test.cc:316]: @asraa sorry I was looking for a TODO here on adding a test?

[the remark this answers is not in the document]
Comment (asraa) [reply test/integration/protocol_integration_test.cc:316]: Ah, I see now! Sorry, I missed this comment. Gotcha, the test triggers the bug and checks that it logs a failure, but I'll add the death test in debug mode.

[the remark this answers is not in the document]
Comment (asraa) [reply test/integration/protocol_integration_test.cc:322]: Done

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**